README.md

London Law
----------

<b>The game still has some bugs, but it can be played now!</b>

London Law is an online multiplayer adaptation of the [classic Scotland Yard
board game](http://www.boardgamegeek.com/boardgame/438/scotland-yard)
([Wikipedia](http://en.wikipedia.org/wiki/Scotland_Yard_board_game)).

This repository is based on a copy of the [last known pre-release](http://www.freelists.org/post/londonlaw-users/030-preview-release)
(0.3pre1) of [London Law](http://pessimization.com/software/londonlaw/) released
by the original author Paul Pelzl and contains additional patches to make this
game work with modern software.

I forked this repo to adapt it to current python and lib versions.

Screenshot:
-----------

[![Screenshot](http://anyc.github.io/londonlaw/images/screenshot_thumb.jpg)](http://anyc.github.io/londonlaw/images/screenshot.jpg)

Dependencies:
-------------
* Python
* [Twisted](https://twistedmatrix.com)
* [wxPython](http://www.wxpython.org/)

The code has been tested with the following versions:
* Python 3.10.9
* Twisted-22.10.0
* wxPython-4.2.0

get started:
------------

Setup is not working, run the server and client manually with the commands below


<h3>For Ubuntu</h3>
<code>sudo apt install python3-wxgtk4.0</code> (only for the client)<br>
<code>sudo apt install python3-twisted</code>

* Run Server:<br>
  python3 londonlaw/london-server.py

* Run Client:<br>
  python3 londonlaw/london-client.py [ip-adress] [player] [password] [gamename]<br>
  (The parameters are optional.)

* Run AI Player:<br>
  python3 londonlaw/aiclients/detective_simple_launcher.py [AiName] [GameNameToJoin]


<h3>For Windows</h3>
Check, if you have Python3 installed:<br>
<code>python --version</code><br>
Python 3.10.9
<br><br>
Install the libraries, you need administration rights:<br>
<code>pip install wxpython</code> (only for the client)<br>
<code>pip install twisted</code>

* Run Server:<br>
  python londonlaw/london-server.py

* Run Client:<br>
  python londonlaw/london-client.py [ip-adress] [player] [password] [gamename]<br>
  (The parameters are optional.)


Changes
-------

1.01.0
   * applied fixes to make it run with current python and lib versions


0.303:
   * ported to python3 by Horst Aldebaran alias Meyer (horald)

0.3.0pre2:

   * wxPython3.0 compatibility by Olly Betts
   * Fix for newer python-twisted by Hans de Goede
   * Accept custom data directory by Mr Bones
   * More small fixes by Mario Kicherer

[0.3.0pre1 by Paul Pelzl](http://www.freelists.org/post/londonlaw-users/030-preview-release):

   * i18n.  Thanks to the efforts of numerous translators, there is
     at least partial support for 11 different languages.
   * AI clients.  Players can request an AI opponent while registering
     for a game.  At the moment there is only a pair of "Rather Dumb"
     AI clients available, meant as a proof of concept.
   * Admin client.  There is a text-mode admin client that can be used
     to delete games, eject players, etc.  It has an online help system,
     which is hopefully enough to explain the usage.

     The admin client won't work without setting up an admin password.
     Create a file called ~/.londonlaw/server/config that looks like
     this:

         [server]
         admin_password: PASSWORD
         game_expiration: 240

     Replace PASSWORD with whatever you want.  The game expiration is
     the number of hours after which stale games should be purged.


doc/ChangeLog
London Law ChangeLog
--------------------------------------------------------------------------------
1.01.0
   2023-03-17        applied fixes to make it run with current python and lib versions

1.00.1
   2021-02-17        Correction to play the game with python3

0.2

   2005-02-15        Announce 0.2.0rc1.

                     Tooltip username labels for the map window are now more
                     visible on OS X.

                     Most wxStaticTextBox instances have been replaced with
                     simple labels.

   2005-02-14        Server will permit clients to disconnect during gameplay
                     and then rejoin the game.  Server saves game data to a
                     database, so it can be shut down and restarted without
                     loss of game state.

                     Clients receive pop-up notification when a player leaves,
                     rejoins, etc.

                     Clients can send chat messages "to all" or "to team".

   2005-02-12        Completed rewrite of all networking code, using the
                     event-driven Twisted framework.  Protocol is now line-based.  
                     Server supports multiple simultaneous games.  Thanks to
                     Conor Davis <conor at fastmail dot fm> for significant 
                     contributions to server code.

                     Client game registration windows now use the more
                     suitable wxListCtrl.


0.1

   2004-11-13        Release 0.1.0.


londonlaw/common/path.py
#  London Law -- a networked manhunting board game
#  Copyright (C) 2003-2004, 2005 Paul Pelzl
#
#  This program is free software; you can redistribute it and/or modify
#  it under the terms of the GNU General Public License, Version 2, as 
#  published by the Free Software Foundation.
#
#  This program is distributed in the hope that it will be useful,
#  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#  GNU General Public License for more details.
#
#  You should have received a copy of the GNU General Public License
#  along with this program; if not, write to the Free Software
#  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA

# path.py
#
# General path and distance-finding algorithms, probably most useful for
# AI programming.
#
# Two pathfinding algorithms are provided: cheapest_path and shortest_path.
# shortest_path is the one to choose when trying to move a pawn from A to B as
# quickly as possible without regard for the tickets spent.  The rest of the
# time, cheapest_path is probably a better choice.

from londonlaw.common import map
import sets



# Cost function which penalizes all tickets (almost) equally.  The small
# cost differences are to provide a slight preference for the
# taxi->bus->underground->black ordering.
def equal_cost(ticket_amounts, ticket_type):
   if ticket_type == 'taxi':
      return 1
   elif ticket_type == 'bus':
      return 1.01
   elif ticket_type == 'underground':
      return 1.02
   else:
      return 1.03

# Same as above, but penalizes black tickets heavily.
def equal_cost_noblack(ticket_amounts, ticket_type):
   if ticket_type == 'taxi':
      return 1
   elif ticket_type == 'bus':
      return 1.01
   elif ticket_type == 'underground':
      return 1.02
   else:
      return 1000000

# Define an unlimited ticket set.
unlimited_tickets = {'taxi' : -1, 'bus' : -1, 'underground' : -1, 'black' : -1}

# Define a ticket set with unlimited non-black tickets.
unlimited_tickets_noblack = {'taxi' : -1, 'bus' : -1, 'underground' : -1, 'black' : 0}



# iterate backward through a 'previous' array to create a path
# in list form (used for cheapest_path and shortest_path)
def _compute_path(source, dest, previous):
   if source == dest:
      return []
   elif previous[dest] == None:
      return None
   else:
      # use the 'previous' array to compute a path in list form
      i = dest
      p = [(dest, previous[dest][1])]
      while previous[i][0] != source:
         p.append((previous[i][0], previous[previous[i][0]][1]))
         i = previous[i][0]
      p.reverse()
      return p 



# Compute the path of least cost from a source vertex to either one destination
# vertex or to all vertices, given an optional cost function that specifies the
# penalty associated with using taxi, bus, underground, and black tickets.
# A ticket set may also be specified, which will be used to eliminate paths that
# violate the ticket constraint.
#
# The cost function takes a set of ticket amounts as the first argument, and a
# ticket type as the second argument.  As the algorithm iterates through longer
# and longer paths, the set of tickets passed to the cost function will change,
# and the AI programmer may take advantage of this to adapt the relative cost
# for each ticket type as tickets become depleted.
# 
# (This is just Dijkstra's algorithm, with the addition of adaptive cost and 
# path pruning to satisfy the ticket constraint.)
#
# If dest=None, returns an array of paths, one for each destination.  If dest
# is a vertex, returns a single path.  (Using dest=None is more efficient than
# separately computing constrained_shortest_path(t, s, d) for all d.)
#
# The path(s) are provided as a list of (destination, transportation) tuples.  When
# no path can be found for a given destination, provides the value None.
#
# Note: since the graph is sparse, this could be done in a more efficient
#       manner... but it seems plenty fast for N=200.
def cheapest_path(source, dest=None, tickets=unlimited_tickets, cost=equal_cost):
   # U is unvisited vertices
   U = set()
   for u in range(1, len(map.locToRoutes)):
      U.add(u)
   path_cost         = [1e100] * len(map.locToRoutes)
   path_cost[source] = 0
   previous          = [None] * len(map.locToRoutes)
   t                 = [None] * len(map.locToRoutes)
   t[source]         = tickets.copy()
   
   while len(U) > 0:
      # extract vertex with minimum path
      min_cost = 1e200
      for u in U:
         if path_cost[u] < min_cost:
            min_cost = path_cost[u]
            min_u    = u
      U.remove(min_u)
      if t[min_u] != None:
         for (v, transports) in map.locToRoutes[min_u]:
            for transport in transports:
               if (path_cost[v] > path_cost[min_u] + cost(t[min_u], transport) and 
               t[min_u][transport] != 0):
                  t[v] = t[min_u].copy()
                  if t[v][transport] > 0:
                     t[v][transport] -= 1
                  path_cost[v] = path_cost[min_u] + cost(t[min_u], transport)
                  previous[v] = (min_u, transport)
   print()
   if dest == None:
      return ([None] + 
      [_compute_path(source, d, previous) for d in range(1, len(map.locToRoutes))])
   else:
      return _compute_path(source, dest, previous)



# Compute the shortest path between two map vertices, given an optional ticket
# constraint.  Uses a breadth-first search.  We always assume that a shorter
# path to a given vertex is better than a longer path, thus limiting the
# algorithmic complexity.
#
# The cost function takes a set of ticket amounts as the first argument, and a
# ticket type as the second argument.  As the algorithm iterates through longer
# and longer paths, the set of tickets passed to the cost function will change,
# and the AI programmer may take advantage of this to adapt the relative cost
# for each ticket type as tickets become depleted.  Specifying a cost function
# will allow the algorithm to judge between two paths of equal length to
# determine which is preferred.
#
# If dest=None, returns an array of paths, one for each destination.  If dest
# is a vertex, returns a single path.  (Using dest=None is more efficient than
# separately computing constrained_shortest_path(t, s, d) for all d.)
#
# The path(s) are provided as a list of (destination, transportation) tuples.  When
# no path can be found for a given destination, provides the value None.
#
def shortest_path(source, dest=None, tickets=unlimited_tickets, cost=equal_cost):
   # V is visited nodes, U is unvisited
   V = set([source])
   U = set()
   for i in range(1, len(map.locToRoutes)):
      U.add(i)
   U.remove(source)

   previous  = [ None ] * len(map.locToRoutes)
   t         = [ None ] * len(map.locToRoutes)
   t[source] = tickets.copy()

   search_vertices = V.copy()
   while (dest == None or dest in U) and len(search_vertices) > 0:
      next_search_vertices = set()
      for visited in search_vertices:
         for d, transports in map.locToRoutes[visited]:
            if d not in V:
               for transport in transports:
                  if t[visited][transport] != 0:
                     if previous[visited] == None:
                        old_cost = 0
                     else:
                        old_cost = previous[visited][2]
                     # If this destination is in next_search_vertices, then we
                     # are considering the case of choosing between two paths
                     # of the same length.  In that situation we must use the
                     # cost function to decide whether to replace the existing path.
                     if (d not in next_search_vertices or cost(t[visited], transport) + 
                     old_cost < previous[d][2]):
                        t[d] = t[visited].copy()
                        if t[d][transport] > 0:
                           t[d][transport] -= 1
                        previous[d] = (visited, transport, 
                              cost(t[visited], transport) + old_cost)
                        next_search_vertices.add(d)
      V.update(next_search_vertices)
      U.difference_update(next_search_vertices)
      search_vertices = next_search_vertices

   if dest == None:
      return ([None] + 
      [_compute_path(source, d, previous) for d in range(1, len(map.locToRoutes))])
   else:
      return _compute_path(source, dest, previous)



# Compute the distance between two map vertices, given optional pathfinding
# function, cost function, and ticket constraint.
#
# If dest=None, returns an array of distances, one for each destination.
# The distance value 'None' is provided when no path can be found which
# satisfies the constraint.
def distance(source, dest=None, pathfinder=shortest_path, 
      cost=equal_cost, tickets=unlimited_tickets):
   def compute_distance(p):
      if p == None:
         return None
      else:
         return len(p)

   paths = pathfinder(source, dest, tickets, cost)
   if dest == None:
      return [compute_distance(p) for p in paths]
   else:
      return compute_distance(paths)



# Given a starting location, number of turns, and optional ticket supply,
# compute the possible locations where a pawn could move.  (This will be slow
# for 'turns' bigger than about 4.)  Returns a Set.
#
# The optional 'eliminate' argument is a list of length 'turns'.  Each element
# of the list is a Set of locations where the pawn cannot move on that turn.
# Thus 'eliminate' could be used to rule out collisions between detectives
# or Mr. X capture scenarios.
#
# If 'force_move' == False, the algorithm assumes that pawns can get stuck due
# to ticket depletion or eliminated locations (as Detectives can), and will
# skip turns as necessary.  If 'force_move' == True, the algorithm will prune
# all scenarios in which the pawn cannot move (useful for determining Mr. X's
# possible destinations).
def possible_destinations(source, turns, tickets=unlimited_tickets, eliminate=None,
      force_move=True):
   if eliminate == None:
      eliminate = [set()] * turns
   # have to convert 'tickets' to a tuple, because otherwise we can't
   # store it in a Set ("dict objects are unhashable")
   tickets_assoc = ()
   for key in tickets:
      tickets_assoc = tickets_assoc + ((key, tickets[key]),)
   loc_sets  = [set([(source, tickets_assoc)])]
   for turn in range(turns):
      loc_sets.append(set())
      for loc, t in loc_sets[turn]:
         is_stuck = 1
         for dest, transports in map.locToRoutes[loc]:
            if dest not in eliminate[turn]:
               for transport in transports:
                  for i in range(len(t)):
                     ticket, amount = t[i]
                     if transport == ticket:
                        if amount != 0:
                           is_stuck = 0
                           if amount > 0:
                              t2 = t[:i] + ((ticket, amount - 1),) + t[i+1:]
                           else:
                              t2 = t
                           loc_sets[turn + 1].add((dest, t2))
         if not force_move and is_stuck:
            # if the pawn is stuck, he just waits in the same location until
            # the next turn
            loc_sets[turn + 1].add((loc, t))

   ret_val = set()
   for loc, t in loc_sets[turns]:
      ret_val.add(loc)
   return ret_val




# Given a starting location and list of tickets spent, compute the possible
# locations where a pawn could have moved.  (This will be slow for a lengthy
# list of tickets.)  Returns a Set.
#
# The optional 'eliminate' argument is a list of the same length as
# 'tickets_spent'.  Each element of the list is a Set of locations where the
# pawn could not have moved on that turn.  Thus 'eliminate' could be used to
# rule out Mr. X capture scenarios.
def possible_locations(source, tickets_spent, eliminate=None):
   if eliminate == None:
      eliminate = [set()] * len(tickets_spent)

   loc_sets = [set([source])]
   for turn in range(len(tickets_spent)):
      loc_sets.append(set())
      for loc in loc_sets[turn]:
         for dest, transports in map.locToRoutes[loc]:
            if tickets_spent[turn] in transports and dest not in eliminate[turn]:
               loc_sets[turn + 1].add(dest)

   return loc_sets[len(tickets_spent)]





londonlaw/guiclient/ScrolledLabel.py
#  London Law -- a networked manhunting board game
#  Copyright (C) 2003-2004, 2005 Paul Pelzl
#
#  This program is free software; you can redistribute it and/or modify
#  it under the terms of the GNU General Public License, Version 2, as 
#  published by the Free Software Foundation.
#
#  This program is distributed in the hope that it will be useful,
#  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#  GNU General Public License for more details.
#
#  You should have received a copy of the GNU General Public License
#  along with this program; if not, write to the Free Software
#  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA




# ScrolledLabel.py
#
# This class handles a wx.StaticText that is managed by a scrolled window.


import wx

class ScrolledLabel(wx.ScrolledWindow):
   def __init__(self, parent, label):
      wx.ScrolledWindow.__init__(self, parent, -1, wx.DefaultPosition, wx.DefaultSize, 
         wx.VSCROLL | wx.SIMPLE_BORDER)
      self.SetBackgroundColour(wx.Colour(200, 200, 200))
      self.SetScrollRate(0, 5)

      # create the text that will be scrolled
      self.text = wx.StaticText(self, -1, label, wx.Point(0,0))

      # use a Sizer to handle geometry
      self.topSizer = wx.BoxSizer(wx.VERTICAL)
      self.topSizer.Add(self.text, 1, wx.EXPAND)
      self.SetSizer(self.topSizer)
      self.topSizer.FitInside(self)

      self.ScrollToEnd()

   def SetText(self, text):
      self.text.SetLabel(text)

      # force the sizer to readjust to the new text, so the scroll bars
      # will properly cover the entire region
      txtsize = self.text.GetSize()
      self.topSizer.SetMinSize(txtsize)
      self.ScrollToEnd()

   def AppendText(self, txt):
      self.text.SetLabel(self.text.GetLabel() + '\n' + txt)

      txtsize = self.text.GetSize()
      self.topSizer.SetMinSize(txtsize)
      self.ScrollToEnd()



   # scroll to the bottom of the text
   def ScrollToEnd(self):
      self.topSizer.FitInside(self)
 #     self.Scroll(0, self.GetVirtualSize()[1]/5)
      self.Scroll(0, 20)




londonlaw/aiclients/x_simple_launcher.py
#!/usr/bin/env python
#
#  London Law -- a networked manhunting board game
#  Copyright (C) 2003-2004, 2005 Paul Pelzl
#
#  This program is free software; you can redistribute it and/or modify
#  it under the terms of the GNU General Public License, Version 2, as 
#  published by the Free Software Foundation.
#
#  This program is distributed in the hope that it will be useful,
#  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#  GNU General Public License for more details.
#
#  You should have received a copy of the GNU General Public License
#  along with this program; if not, write to the Free Software
#  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA

# Command-line launcher for x_simple AI.  Use it like
# $ python x_simple_launcher.py "gameroom name" PORT
# (where PORT is optional, defaulting to the obvious choice)
#
# The gameroom name should be encoded using UTF-8.


from twisted.internet import reactor
from twisted.python import log
import sys, os


# add the parent directories to PYTHONPATH
sys.path.append(os.getcwd())
parent = os.path.split(os.path.abspath(os.getcwd()))[0]
root   = os.path.split(os.path.abspath(parent))[0]
sys.path.append(parent)
sys.path.append(root)
from londonlaw.aiclients import x_simple
from londonlaw.common.protocol import LLAW_PORT


log.startLogging(sys.stderr, 0)
if len(sys.argv) > 3:
   port = int(sys.argv[3])
else:
   port = LLAW_PORT
log.msg("Using server at localhost port " + str(port))
log.msg("Sending simple AI client \"" + sys.argv[1] + "\" to game room \"" + sys.argv[2] + "\"")

f = x_simple.XSimpleAIFactory(sys.argv[1], sys.argv[2])
reactor.connectTCP("localhost", port, f)
reactor.run()





londonlaw/aiclients/detective_simple_launcher.py
#!/usr/bin/env python
#
#  London Law -- a networked manhunting board game
#  Copyright (C) 2003-2004, 2005 Paul Pelzl
#
#  This program is free software; you can redistribute it and/or modify
#  it under the terms of the GNU General Public License, Version 2, as 
#  published by the Free Software Foundation.
#
#  This program is distributed in the hope that it will be useful,
#  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#  GNU General Public License for more details.
#
#  You should have received a copy of the GNU General Public License
#  along with this program; if not, write to the Free Software
#  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA

# Command-line launcher for detective_simple AI.  Use it like
# $ python detective_simple_launcher.py "gameroom name" PORT
# (where PORT is optional, defaulting to the obvious choice)
#
# The gameroom name should be encoded using UTF-8.


from twisted.internet import reactor
from twisted.python import log
import sys, os


# add the parent directories to PYTHONPATH
sys.path.append(os.getcwd())
parent = os.path.split(os.path.abspath(os.getcwd()))[0]
root   = os.path.split(os.path.abspath(parent))[0]
sys.path.append(parent)
sys.path.append(root)
from londonlaw.aiclients import detective_simple
from londonlaw.common.protocol import LLAW_PORT


log.startLogging(sys.stderr, 0)
if len(sys.argv) > 3:
   port = int(sys.argv[3])
else:
   port = LLAW_PORT
log.msg("Using server at localhost port " + str(port))
log.msg("Sending simple AI client \"" + sys.argv[1] + "\" to game room \"" + sys.argv[2] + "\"")

f = detective_simple.DetectiveSimpleAIFactory(sys.argv[1], sys.argv[2])
reactor.connectTCP("localhost", port, f)
reactor.run()





londonlaw/aiclients/base.py
#  London Law -- a networked manhunting board game
#  Copyright (C) 2003-2004, 2005 Paul Pelzl
#
#  This program is free software; you can redistribute it and/or modify
#  it under the terms of the GNU General Public License, Version 2, as 
#  published by the Free Software Foundation.
#
#  This program is distributed in the hope that it will be useful,
#  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#  GNU General Public License for more details.
#
#  You should have received a copy of the GNU General Public License
#  along with this program; if not, write to the Free Software
#  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA



from twisted.internet import protocol, reactor
from twisted.protocols import basic
from twisted.python import log
import re, shlex, sys, sets
from londonlaw.common.protocol import *
from londonlaw.common.Pawn import *
from londonlaw.common import util, path, map


class BaseAIProtocolError(Exception):
   pass


# Base class for AI players.  Knows how to communicate with the server,
# keeps track of pawn information and game history, has some basic
# pathfinding/distance algorithms.
class BaseAIProtocol(basic.LineOnlyReceiver):

   def __init__(self):
      self._state              = "init"
      self._tagIndex           = 1
      self._pawns              = {}
      self._myPawns            = {}
      self._history            = [{} for i in range(25)]
      self._turnNum            = 0
      self._lastMover          = None
      self._lastXSurfacingTurn = None
      self._players            = set()
      self._gameStatus         = None


   def connectionLost(self, reason):
      #reactor.stop()
      print ("AI: connection lost")


   def connectionMade(self):
      self._username = self.factory.username
      self._password = self.factory.password
      self._gameroom = self.factory.gameroom
      self._team     = self.factory.team
      self._state    = "protocol"
      self.sendTokens(self.genTag(), "protocol", PROTOCOL_VERSION)


   # called when this AI needs to make a move
   def doTurn(self, pawnName):
      raise NotImplementedError("doTurn must be implemented in derived classes.")


   # Find sets of locations where Mr. X safely move on the next turn.
   # Returns a list of Sets: the first Set is locations at least distance
   # 1 from all detectives, the second Set is locations at least distance
   # 2, etc. (up to the maximum map distance).  Note that moves at distance
   # 1 are UNsafe.
   def safeMoves(self, loc=None):
      if loc == None:
         loc = self._pawns['X'].getLocation()
      if loc == -1:
         raise BaseAIProtocolError("Cannot determine safe X moves when X location is unknown")
       
      dets      = ['Red', 'Yellow', 'Green', 'Blue', 'Black']
      detLocs   = [self._pawns[d].getLocation() for d in dets]
      allMoves  = path.possible_destinations(loc, 1, eliminate=[set(detLocs)])
      distances = [path.distance(self._pawns[d].getLocation(),
            tickets=self._pawns[d]._tickets) for d in dets]

      def findSafe(threshold):
         safe = allMoves.copy()
         for dest in allMoves:
            for dist in distances:
               if dist[dest] < threshold:
                  safe.remove(dest)
                  break
         return safe

      return [findSafe(threshold) for threshold in range(1, map.MAX_DISTANCE + 1)]


   def genTag(self):
      if self._tagIndex > 99999:
         self._tagIndex = 0
      st = repr(self._tagIndex)
      self._tagIndex += 1
      self._waitTag = "#" + "0"*(5 - len(st)) + st
      return self._waitTag

   
   def lineReceived(self, line):
      try:
#         tokens = shlex.split(line)
         tokens = shlex.split(line.decode("utf-8"))
         if len(tokens) > 1:
            tag       = tokens[0]
            response  = tokens[1].lower()
            data      = tokens[2:]

            f = getattr(self, "".join(("response_", response, "_", self._state)), None)
            if f is None:
               # Try a catch-all method for the command
               f = getattr(self, "".join(("response_", response, "_default")), None)
            if f is None:
               log.msg("Received unhandled server message (tried default): \"" + line + "\" state = \"" + self._state + "\"")
               return
            
            f(tag, data)
         else:
            log.msg("Received unhandled server message (too few args): \"" + line + "\" state = \"" + self._state + "\"")

      except AttributeError:
         log.msg(str(AttributeError))
         #log.msg("tokens = " + str(tokens))
         log.msg("Received unhandled server message: \"" + str(line) + "\" state = \"" + self._state + "\"")


   def makeMove(self, data):
      self._state = "trymove"
      log.msg("move data = " + str(data))
      if len(data) == 3:
         self.sendTokens(self.genTag(), "move", data[0], data[1], data[2])
      elif len(data) == 5:
         self.sendTokens(self.genTag(), "doublemove", 
               data[0], data[1], data[2], data[3], data[4])
      else:
         raise BaseAIProtocolError("makeMove() received illegal data length; data = " + str(data))


   def response_bad_default(self, tag, args):
      log.msg("received 'bad' response, state = " + self._state + ", reason = " + str(args))
      self.transport.loseConnection()


   def response_chatall_default(self, tag, args):
      pass


   def response_chatteam_default(self, tag, args):
      pass


   def response_ejected_default(self, tag, data):
      log.msg("received 'ejected' response")
      self.transport.loseConnection()


   def response_gameover_default(self, tag, args):
      self.transport.loseConnection()


   def response_gameinfo_default(self, tag, args):
      if len(args) >= 2:
         try:
#            name   = args[0].decode("utf-8")
            name   = args[0]
            status = args[1]
            if name == self._gameroom:
               self._gameStatus = status
         except:
            pass


   def response_gamestart_default(self, tag, args):
      self._state = "playing"


   def response_history_playing(self, tag, args):
      moves = [moveStr.split() for moveStr in args[1:]]
      for move in moves:
         if args[0] == '0':
            self._history[0][move[0]] = int(move[1])
         else:
            self._history[int(args[0])][move[0]] = (int(move[1]), move[2])
            if move[0] == 'X' and int(move[1]) != -1:
               self._lastXSurfacingTurn = int(args[0])


   def response_move_playing(self, tag, args):
      self._pawns[args[0]].setLocation(int(args[1]))
      self._pawns[args[0]].removeTicket(args[2])
      self._history[self._turnNum][args[0]] = (int(args[1]), args[2])
      if args[0] == "X" and int(args[1]) != -1:
         self._lastXSurfacingTurn = self._turnNum
      if args[0] == self._lastMover:
         self._pawns[args[0]].removeTicket("double")
      self._lastMover = args[0]


   def response_no_default(self, tag, args):
      log.msg("received 'no' response, state = " + self._state + ", reason = " + str(args))
      self.transport.loseConnection()


   def response_no_protocol(self, tag, args):
      self.transport.loseConnection()


   def response_no_trysetteam(self, tag, args):
      self.sendChat("I was unable to select team \"" + self._team + 
            "\".  You need to make room for me on that team!", "all")
      self.sendChat("Logging out.", "all")
      self.transport.loseConnection()


   def response_ok_trylistgames(self, tag, args):
      if self._gameStatus == GAMESTATUS_NEW:
         self._state = "tryjoin"
         self.sendTokens(self.genTag(), "join", self._gameroom)
      elif self._gameStatus == GAMESTATUS_INPROGRESS:
         self._state = "tryrejoin"
         self.sendTokens(self.genTag(), "join", self._gameroom)
      elif self._gameStatus == None:
         log.msg("Specified game not found; exiting.")
         self.transport.loseConnection()
      else:
         log.msg("Cannot join completed game; exiting.")
         self.transport.loseConnection()


   def response_ok_login(self, tag, args):
      self._state = "trylistgames"
      self.sendTokens(self.genTag(), "listgames")


   def response_ok_playing(self, tag, args):
      pass


   def response_ok_protocol(self, tag, args):
      self._state = "login"
      self.sendTokens(self.genTag(), "login", self._username, self._password)


   def response_ok_tryjoin(self, tag, args):
      self._state = "trylistplayers"
      self.sendTokens(self.genTag(), "listplayers")


   def response_ok_trylistplayers(self, tag, args):
      self._state = "trysetteam"
      self.sendTokens(self.genTag(), "setteam", self._team)


   def response_ok_trymove(self, tag, data):
      self._state = "playing"


   def response_ok_tryrejoin(self, tag, args):
      log.msg("rejoined an in-progress game")
      self._state = "playing"


   def response_ok_trysetteam(self, tag, args):
      self._state = "tryvote"
      self.sendTokens(self.genTag(), "votestart", "True")


   def response_ok_tryvote(self, tag, args):
      self._state = "ready"


   def response_pawninfo_playing(self, tag, args):
      p = Pawn(args[0])
#      p.setPlayer(args[1].decode("utf-8"))
      p.setPlayer(args[1])
      p.setLocation(int(args[2]))
      p.setTicketAmount("taxi", int(args[3]))
      p.setTicketAmount("bus", int(args[4]))
      p.setTicketAmount("underground", int(args[5]))
      p.setTicketAmount("black", int(args[6]))
      p.setTicketAmount("double", int(args[7]))

      self._pawns[p.getName()] = p
      if p.getPlayer() == self._username:
         self._myPawns[p.getName()] = p
      
      if p.getName() == "X":
         self._history[0]['X'] = p.getLocation()
   

   def response_stuck_playing(self, tag, args):
      pass


   def response_turn_playing(self, tag, args):
      if args[0] in self._myPawns.keys():
         self.doTurn(args[0])


   def response_turnnum_playing(self, tag, args):
      self._turnNum = int(args[0])


   def response_playerinfo_default(self, tag, args):
#      self._players.add(args[0].decode("utf-8"))
      self._players.add(args[0])


   def response_playerleave_default(self, tag, args):
      pass


   def response_playerleave_ready(self, tag, args):
      username = args[0].decode("utf-8")
      if username in self._players:
         self._players.remove(username)
      if len(self._players) == 1:
         # if this bot is the last player, leave the gameroom
         self.transport.loseConnection()


   def response_playerleave_trysetteam(self, tag, args):
      self.response_playerleave_ready(tag, args)


   def response_playerleave_votestart(self, tag, args):
      self.response_playerleave_ready(tag, args)


   def response_rejoin_default(self, tag, args):
      pass


   def sendChat(self, text, sendTo):
      if sendTo == "all":
         self.sendTokens(self.genTag(), "chatall", text.encode("utf-8"))
      elif sendTo == "team":
         self.sendTokens(self.genTag(), "chatteam", text.encode("utf-8"))


   def sendTokens(self, *tokens):

      s = util.join_tokens(*tokens)
      print(str(s))
      self.sendLine(s.encode("utf-8"))



class BaseAIFactory(protocol.ClientFactory):
   protocol = BaseAIProtocol

   def __init__(self, username, password, gameroom, team):
      self.username = username
      self.password = password
      self.gameroom = gameroom
      self.team     = team

   def clientConnectionFailed(self, connector, reason):
      log.msg("Failed to connect to specified server.")
      reactor.stop()



